pkg/kubelet/dockershim/docker_sandbox.go

```diff
@@ -203,38 +205,69 @@ func (ds *dockerService) StopPodSandbox(podSandboxID string) error {
 			errList = append(errList, err)
 		}
 	}
-	if err := ds.client.StopContainer(podSandboxID, defaultSandboxGracePeriod); err != nil {
-		glog.Errorf("Failed to stop sandbox %q: %v", podSandboxID, err)
-		// Do not return error if the container does not exist
-		if !dockertools.IsContainerNotFoundError(err) {
-			errList = append(errList, err)
+
+	// Get all the containers in the PodSandbox.
+	containers, err := ds.getContainersInPodSandbox(podSandboxID)
+	if err != nil {
+		errList = append(errList, err)
+	}
+
+	// Stop all containers in the sandbox.
+	for _, container := range containers {
+		if container.State == "running" {
+			wg.Add(1)
+			go func(containerID string) {
+				defer wg.Done()
+				if err := ds.StopContainer(containerID, int64(defaultSandboxGracePeriod)); err != nil && !dockertools.IsContainerNotFoundError(err) {
+					errList = append(errList, err)
+				}
+			}(container.ID)
 		}
 	}
+	wg.Wait()
+
+	// Stop sandbox container only when there are no running regular containers.
```

Comment:
so some extra comments

1) if the first getContainersInPodSandbox() fails, why don't we return right away?  Is the logic, just try to clean up as much as we can this time around?

2) I think my argument was simply that you could do something like

```go
if len(errList) == 0 {
   ds.client.StopContainer(podSandboxID, defaultSandboxGracePeriod) //obviously with the full check as it exists
}
```

i.e. we trust the list call above and we also know if anything failed above, so we only remove the sandbox container if nothing failed (and hence everything should be gone).  It seems overkill (though perhaps I'm wrong)

As an aside my other Q is, is this correct?

I can imagine a negative scenario where some resource fails to be removed in the old code, so we get a leak, but since we can remove the pod sandbox container at least we can recreate a pod with the same name.

is it better to have a sandbox container that can not be removed because of some underlying fault OR to leak resources that we wont really be accounting for and remove the sandox container.